client/executor/common/src/sandbox/wasmi_backend.rs

```diff
@@ -0,0 +1,341 @@
+	},
+	util::{checked_range, MemoryTransfer},
+};
+
+environmental::environmental!(SandboxContextStore: trait SandboxContext);
+
+/// Construct trap error from specified message
+fn trap(msg: &'static str) -> Trap {
+	TrapKind::Host(Box::new(Error::Other(msg.into()))).into()
+}
+
+impl ImportResolver for Imports {
+	fn resolve_func(
+		&self,
+		module_name: &str,
+		field_name: &str,
+		signature: &::wasmi::Signature,
+	) -> std::result::Result<wasmi::FuncRef, wasmi::Error> {
+		let idx = self.func_by_name(module_name, field_name).ok_or_else(|| {
+			wasmi::Error::Instantiation(format!("Export {}:{} not found", module_name, field_name))
+		})?;
+
+		Ok(wasmi::FuncInstance::alloc_host(signature.clone(), idx.0))
+	}
+
+	fn resolve_memory(
+		&self,
+		module_name: &str,
+		field_name: &str,
+		_memory_type: &::wasmi::MemoryDescriptor,
+	) -> std::result::Result<wasmi::MemoryRef, wasmi::Error> {
+		let mem = self.memory_by_name(module_name, field_name).ok_or_else(|| {
+			wasmi::Error::Instantiation(format!("Export {}:{} not found", module_name, field_name))
+		})?;
+
+		let wrapper = mem.as_wasmi().ok_or_else(|| {
+			wasmi::Error::Instantiation(format!(
+				"Unsupported non-wasmi export {}:{}",
+				module_name, field_name
+			))
+		})?;
+
+		// Here we use inner memory reference only to resolve
+		// the imports without accessing the memory contents.
+		let mem = unsafe { wrapper.clone_inner() };
+
+		Ok(mem)
+	}
+
+	fn resolve_global(
+		&self,
+		module_name: &str,
+		field_name: &str,
+		_global_type: &::wasmi::GlobalDescriptor,
+	) -> std::result::Result<wasmi::GlobalRef, wasmi::Error> {
+		Err(wasmi::Error::Instantiation(format!("Export {}:{} not found", module_name, field_name)))
+	}
+
+	fn resolve_table(
+		&self,
+		module_name: &str,
+		field_name: &str,
+		_table_type: &::wasmi::TableDescriptor,
+	) -> std::result::Result<wasmi::TableRef, wasmi::Error> {
+		Err(wasmi::Error::Instantiation(format!("Export {}:{} not found", module_name, field_name)))
+	}
+}
+
+pub fn new_memory(initial: u32, maximum: Option<u32>) -> crate::error::Result<Memory> {
+	let memory = Memory::Wasmi(MemoryWrapper::new(MemoryInstance::alloc(
+		Pages(initial as usize),
+		maximum.map(|m| Pages(m as usize)),
+	)?));
+
+	Ok(memory)
+}
+
+/// Wasmi provides direct access to its memory using slices.
+///
+/// This wrapper limits the scope where the slice can be taken to
+#[derive(Debug, Clone)]
+pub struct MemoryWrapper(::wasmi::MemoryRef);
+
+impl MemoryWrapper {
+	/// Take ownership of the memory region and return a wrapper object
+	pub fn new(memory: ::wasmi::MemoryRef) -> Self {
+		Self(memory)
+	}
+
+	/// Clone the underlying memory object
+	///
+	/// # Safety
+	///
+	/// The sole purpose of `MemoryRef` is to protect the memory from uncontrolled
+	/// access. By returning the memory object "as is" we bypass all of the checks.
+	///
+	/// Intended to use only during module initialization.
+	pub unsafe fn clone_inner(&self) -> ::wasmi::MemoryRef {
```

Comment:
```suggestion
	pub unsafe fn clone_inner(&self) -> wasmi::MemoryRef {
```

I don't see why this is "unsafe". Is the memory getting corrupted if you clone it? I don't think so?

Reply:
The `clone_inner` is also used in this file, so we could also just access `wrapper.0` and drop this function entirely.